includes/abstracts/llms-abstract-controller-user-engagements.php

```diff
@@ -0,0 +1,234 @@
+ *
+ * @since [version]
+ */
+abstract class LLMS_Abstract_Controller_User_Engagements
+	implements LLMS_Interface_Case, LLMS_Interface_User_Engagement_Type {
+
+	use LLMS_Trait_User_Engagement_Type;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function __construct() {
+
+		add_action( 'init', array( $this, 'maybe_handle_awarded_engagement_sync_actions' ) );
+	}
+
+	/**
+	 * Delete an awarded user engagement.
+	 *
+	 * @since 3.18.0
+	 * @since [version] Permanently delete user engagement via wp_delete_post().
+	 *              Refactored from LLMS_Controller_Achievements::delete() and LLMS_Controller_Certificates::delete().
+	 *
+	 * @param int $post_id WP Post ID of the awarded engagement.
+	 * @return void
+	 */
+	protected function delete( $post_id ) {
+
+		if ( ! is_admin() ) {
+			return;
+		}
+
+		wp_delete_post( $post_id, true );
+	}
+
+	/**
+	 * Handle awarded engagement sync actions.
+	 *
+	 * Errors are added to {@see LLMS_Admin_Metabox::add_error()} to be displayed as an admin notice
+	 * and also returned for unit tests.
+	 *
+	 * If the sync is successful, the {@see llms_redirect_and_exit()} function is called and this method does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @return null|WP_Error
+	 */
+	public function maybe_handle_awarded_engagement_sync_actions() {
+
+		// Validate action.
+		// Invalid actions return a WP_Error for testing purposes and are not displayed to the user.
+		$actions = array(
+			'sync_one'  => "sync_awarded_$this->engagement_type",
+			'sync_many' => "sync_awarded_{$this->engagement_type}s",
+		);
+		if ( ! isset( $_GET['action'] ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-missing-action",
+				__( 'Sorry, you have not provided any actions.', 'lifterlms' )
+			);
+		} elseif ( ! in_array( $_GET['action'], $actions ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-invalid-action",
+				__( 'You\'re trying to perform an invalid action.', 'lifterlms' )
+			);
+		}
+
+		// Verify nonce.
+		if ( ! llms_verify_nonce(
+			"_llms_{$this->engagement_type}_sync_actions_nonce",
+			"llms-$this->engagement_type-sync-actions",
+			'GET'
+		) ) {
+			$result = new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-invalid-nonce",
+				sprintf(
+					/* translators: 1: plural lowercase awarded engagement type, 2: plural uppercase first letter awarded engagement type */
+					__( 'Sorry, you are not allowed to sync %1$s.', 'lifterlms' ),
+					$this->get_engagement_type_name( self::PLURAL, self::CASE_LOWER, self::AWARDED ),
+					$this->get_engagement_type_name( self::PLURAL, self::CASE_UPPER_FIRST, self::AWARDED )
+				)
+			);
+			( new LLMS_Meta_Box_Award_Engagement_Submit() )->add_error( $result );
+
+			return $result;
+		}
+
+		$engagement_id = llms_filter_input( INPUT_GET, 'post', FILTER_SANITIZE_NUMBER_INT );
+
```

Comment:
I'd like to see the two `empty( $engagement_id )` checks / errors combined into a single error here. They're basically both the same except for the post type name. I would suggest moving the generation of the error message into a private helper function, something like `get_missing_post_id_error_message( $action )` or something like that... Seems like a lot of redundant code to me is all.

Reply:
@thomasplevy,
With the replacement of `LLMS_Trait_User_Engagement_Type::get_engagement_type_name()` with `LLMS_Abstract_Controller_User_Engagements::get_text()`, generating the error message is much more compact and easier to follow. There's still a fair bit of indentation going on.
```php
if ( $action === $actions['sync_one'] ) {
    if ( empty( $engagement_id ) ) {
        $result = new WP_Error(
            "llms-sync-missing-awarded-$this->engagement_type-id",
            $this->get_text( self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID )
        );
    } else {
        $result = $this->sync_awarded_engagement( $engagement_id );
    }
} elseif ( $action === $actions['sync_many'] ) {
    if ( empty( $engagement_id ) ) {
        $result = new WP_Error(
            "llms-sync-missing-{$this->engagement_type}-template-id",
            $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID )
        );
    } else {
        $result = $this->sync_awarded_engagements( $engagement_id );
    }
} else {
    $result = null;
}
```

How about this? Is it any better?
```php
$is_syncing_one = $action === $actions['sync_one'];
if ( empty( $engagement_id ) ) {
    if ( $is_syncing_one ) {
        $code    = "llms-sync-missing-awarded-$this->engagement_type-id";
        $message = $this->get_text( self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID );
    } else {
        $code    = "llms-sync-missing-{$this->engagement_type}-template-id";
        $message = $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID );
    }
    $result = new WP_Error( $code, $message );
} else {
    if ( $is_syncing_one ) {
        $result = $this->sync_awarded_engagement( $engagement_id );
    } else {
        $result = $this->sync_awarded_engagements( $engagement_id );
    }
}
```

Reply:
I committed the last option.